sklearn/ensemble/_hist_gradient_boosting/splitting.pyx

```diff
@@ -556,7 +565,7 @@ cdef class Splitter:
             Y_DTYPE_C sum_gradient_right
             Y_DTYPE_C negative_loss_current_node
             Y_DTYPE_C gain
-            unsigned int start = self.n_bins_non_missing[feature_idx] - 2
+            unsigned int start = self.n_bins_non_missing[feature_idx] - 1
```

Comment:
By using `-1`, you are counting the missing values bin as part of the right node at the first iteration of the loop.

I still need to convince myself that this is the right fix (and that we don't need something else too), but reverting it to `-2` seems to fix the problem.